packages/toolpad-app/middleware.ts

```diff
@@ -1,24 +1,27 @@
-import { NextRequest, NextResponse } from 'next/server';
-import config from './src/server/config';
+import { NextRequest, NextResponse, URLPattern } from 'next/server';
+import { checkBasicAuthHeader } from './src/server/basicAuth';
 
-const BASIC_AUTH_WHITELIST = new Set(['/health-check']);
+const BASIC_AUTH_WHITELIST = [
+  // Healthcheck must always be public
+  new URLPattern({ pathname: '/health-check' }),
 
-export function middleware(req: NextRequest) {
-  const { pathname } = new URL(req.url);
+  // Static content is public
+  new URLPattern({ pathname: '/_next/static/*' }),
+
+  // Apps must be able to be public
+  // These urls will handle their own basic auth when the corresponding Toolpad app is not public
+  new URLPattern({ pathname: '/deploy/*' }),
+  new URLPattern({ pathname: '/api/data/:appId/:version/:queryId' }),
+];
 
-  if (!config.basicAuthUser || BASIC_AUTH_WHITELIST.has(pathname)) {
+export function middleware(req: NextRequest) {
+  const isWhitelisted = BASIC_AUTH_WHITELIST.some((pattern) => pattern.test(req.url));
+  if (isWhitelisted) {
     return NextResponse.next();
   }
 
-  const basicAuth = req.headers.get('authorization');
-
-  if (basicAuth) {
-    const auth = basicAuth.split(' ')[1];
-    const [user, pwd] = atob(auth).toString().split(':');
-
-    if (user === config.basicAuthUser && pwd === config.basicAuthPassword) {
-      return NextResponse.next();
-    }
+  if (checkBasicAuthHeader(req.headers.get('authorization'))) {
+    return NextResponse.next();
   }
 
   return new Response(null, {
```

Comment:
we might need to change this middleware according to some changes here:
https://nextjs.org/docs/messages/middleware-upgrade-guide

basically using `rewrite/redirect` for the response, and maybe we can use the `config` export for the whitelisting?

example here: https://github.com/vercel/examples/tree/main/edge-functions/basic-auth-password

Reply:
Empty body responses are still be allowed. This code has already been passed through the upgrade guide. (See this example where an empty body response is created: https://nextjs.org/docs/messages/middleware-upgrade-guide#after-1)

Reply:
ok so empty body still works, my bad
about the config sounds good too